cdap-system-metrics/src/main/java/io/cdap/cdap/metrics/publisher/MetricsWritersMetricsPublisher.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.metrics.publisher;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.metrics.MetricValues;
+import io.cdap.cdap.api.metrics.MetricsPublisher;
+import io.cdap.cdap.api.metrics.MetricsWriter;
+import io.cdap.cdap.api.metrics.NoopMetricsContext;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.metrics.process.DefaultMetricsWriterContext;
+import io.cdap.cdap.metrics.process.loader.MetricsWriterProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * A {@link MetricsPublisher} that writes the published metrics to multiple {@link MetricsWriter}s
+ */
+public class MetricsWritersMetricsPublisher extends AbstractMetricsPublisher {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsWritersMetricsPublisher.class);
+  private final Map<String, MetricsWriter> metricsWriters;
+
+  @Inject
+  public MetricsWritersMetricsPublisher(MetricsWriterProvider writerProvider, CConfiguration cConf) {
+    this.metricsWriters = writerProvider.loadMetricsWriters();
+    initializeMetricWriters(this.metricsWriters, cConf);
+  }
+
+  private static void initializeMetricWriters(Map<String, MetricsWriter> metricsWriters, CConfiguration cConf) {
```

Comment:
better to initialize in an initialize() method in publisher. Avoid static methods that work on the state variables.

Reply:
Added an initialize method in Publisher interface. Added a protected variable in abstract publisher to keep track of whether publisher has been initialized. Added guard conditions in publish methods to throw exceptions if publisher isn't initialized at the time publishing.